### PR TITLE
362 full import

### LINKED ISSUE #362: Support full import statements
Things like this aren't supported:

```
import sheen.models.Discriminator
import scipy.io.wavfile.read
```

Currently we are forced to write:

```
from sheen.models import Discriminator
from scipy.io.wavfile import read
```

### PR DESCRIPTION
**Pull request by wuhu**
## Description

Fixes #362

### CODE REVIEW

**Comment by sjrl** [comment]
It looks like the failing test has to do with a convention in python 3.9. Here is the error message
```python
>       assert dump == read_dump('othermodule_full_makeclasstransform_with_transforms')
E       AssertionError: assert 'from padl im...rm(a, b, c)\n' == 'from padl im...rm(a, b, c)\n'
E         Skipping 312 identical leading characters in diff, use -v to show
E         -    return (tests_material_transforms_in_module_CONST + k(y))
E         ?           -                                                -
E         +    return tests_material_transforms_in_module_CONST + k(y)
```

File: padl/dumptools/symfinder.py [hunk 1]
```diff
@@ -342,7 +339,18 @@ def deparse(self):
             source = self._result._source
         except AttributeError:
             source = self.source
-        return f'{self.var_name} = {ast_utils.get_source_segment(source, self._result.value)}'
+        return ast_utils.get_source_segment(source, self._result)
+
+
+class _SetAttribute(ast.NodeVisitor):
```
Comment (sjrl) [inline padl/dumptools/symfinder.py:345]: I know this is a simple class, but I would appreciate a short docstring explaining what's going on.

File: padl/dumptools/inspector.py [hunk 2]
```diff
@@ -50,30 +50,41 @@ def _determine_scope(self, caller_frameinfo: inspect.FrameInfo,
...
+
+        return _get_scope_from_frame(caller_frameinfo.frame, drop_n)
 
     @property
     def module(self):
         """The calling module. """
         return self.scope.module
 
 
+def _get_scope_from_frame(frame, drop_n):
```
Comment (sjrl) [inline padl/dumptools/inspector.py:62]: Could you add a docstring for this function?

File: padl/dumptools/var2mod.py [hunk 3]
```diff
@@ -140,10 +157,45 @@ def visit_Name(self, node):
...
+        Vars(globals={ScopedName(name='x', scope=None, n=0)}, locals=set())
         """
-        if (node.id, 0) not in self.locals:
-            self.globals.add((node.id, 0))
+        scope = getattr(node, '_scope', None)
+        name = ScopedName(node.id, scope, 0)
+        if not self.in_locals(name):
+            self.globals.add(name)
+
+    def visit_Attribute(self, node):
```
Comment (sjrl) [inline padl/dumptools/var2mod.py:167]: Should `Attribute` be lower case here in the function name?

**Review by sjrl — APPROVED** [review]
Looks good!

File: padl/dumptools/symfinder.py (hunk 1 above)
Comment (wuhu) [reply padl/dumptools/symfinder.py:345]: I'll add

File: padl/dumptools/inspector.py (hunk 2 above)
Comment (wuhu) [reply padl/dumptools/inspector.py:62]: yes

File: padl/dumptools/var2mod.py (hunk 3 above)
Comment (sjrl) [reply padl/dumptools/var2mod.py:167]: Thanks for the explanation

**Comment by jasonkhadka** [comment]
tests still failing for 3.9 here.

File: padl/dumptools/symfinder.py [hunk 4]
```diff
@@ -342,7 +339,32 @@ def deparse(self):
...
+    """
+
+    def __init__(self, attr, value):
+        self.attr = attr
+        self.value = value
+
+    def generic_visit(self, node):
+        setattr(node, self.attr, self.value)
+        super().generic_visit(node)
+
```
Comment (jasonkhadka) [inline padl/dumptools/symfinder.py:367]: extra line here.

File: padl/dumptools/symfinder.py [hunk 5]
```diff
@@ -606,6 +633,9 @@ def f(x):
     def __hash__(self):
         return hash((self.name, self.scope, self.n))
 
+    def __eq__(self, other):
+        return (self.name, self.scope, self.n) == (other.name, other.scope, other.n)
+
 
 def find_in_scope(name: ScopedName):
```
Comment (jasonkhadka) [inline padl/dumptools/symfinder.py:640]: docstring needs udpate.

**Comment by wuhu** [comment]
@jasonkhadka I'll merge with failing windows tests and create an issue for it. We can decide later whether to drop support.

**Pull request merged**